crypto/crypto.go

```diff
@@ -66,31 +69,52 @@ func NewKeccakState() KeccakState {
 
 // HashData hashes the provided data using the KeccakState and returns a 32 byte hash
 func HashData(kh KeccakState, data []byte) (h common.Hash) {
+	if hash, ok := keccakState256Cache.HasGet(nil, data); ok {
+		return common.BytesToHash(hash)
+	}
 	kh.Reset()
 	kh.Write(data)
 	kh.Read(h[:])
+	keccakState256Cache.Set(data, h.Bytes())
 	return h
 }
 
 // Keccak256 calculates and returns the Keccak256 hash of the input data.
 func Keccak256(data ...[]byte) []byte {
+	if len(data) == 1 {
```

Comment:
could you add some comments why we need some special logic for length == 1

Reply:
Just confirmed with the original contributor, the `length==1` case is the most common case which means that the method only accept one byte array as input. 
And by processing on the one input case, it is also reduce the complexity of the logic here since there is no need to concern the concatenation of bytearrays and hash cache calculations.

Reply:
I will add brief comments in the code.